command/agent/auth/auth.go

```diff
@@ -71,18 +71,36 @@ func backoffOrQuit(ctx context.Context, backoff time.Duration) {
 	}
 }
 
+// Run is a convienence method to call RunWithConnectionTimeout with the default value
+// of 0 connTimeout, which indicates no maximum attempts.
 func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) {
+	ah.RunWithConnectionTimeout(ctx, am, 0)
+}
+
+// Run is a long running process that authenticates with Vault, retrieves an
+// access token, and passes that token to it's output channels to be consumed
+// down the pipeline. The MaxAttempts variable determines how many attempts can
+// be made to a non-responsive Vault instance. A value of 0 for
+// connTimeout means there is no limit to the connection attemts
+// (Vault Agent will not abort due to no connection being established).
+func (ah *AuthHandler) RunWithConnectionTimeout(ctx context.Context, am AuthMethod, connTimeout time.Duration) {
 	if am == nil {
 		panic("nil auth method")
 	}
 
+	var connTimer *time.Timer
 	ah.logger.Info("starting auth handler")
 	defer func() {
 		am.Shutdown()
 		close(ah.OutputCh)
 		close(ah.DoneCh)
 		close(ah.TemplateTokenCh)
 		ah.logger.Info("auth handler stopped")
+		if connTimer != nil {
+			if !connTimer.Stop() {
```

Comment:
I see this example in the Go docs but I don't really understand it, and it feels like we could hang in the channel receive?